# <b> Peer-graded Assignment: Building a Regression Model in Keras </b>

## Part A

In [1]:
import pandas as pd
import numpy as np
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column
n_cols = predictors.shape[1]

In [2]:
import keras
from keras.models import Sequential
from keras.layers import Dense

# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
mselist=[]
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    mse=mean_squared_error(y_test, y_pred)
    mselist.append(mse)
    print('Mean squared error:',mse )
mselist

Mean squared error: 126.29349122419045
Mean squared error: 96.76143297806836
Mean squared error: 147.78868716177516
Mean squared error: 296.0136465608284
Mean squared error: 91.91169310247287
Mean squared error: 127.28017751071708
Mean squared error: 111.08590586565504
Mean squared error: 1045.128800130171
Mean squared error: 813.8786820929685
Mean squared error: 179.3520821927464
Mean squared error: 1183.900762301564
Mean squared error: 177.73368351466274
Mean squared error: 105.84584134997014
Mean squared error: 152.99098443251918
Mean squared error: 1266.5163162432736
Mean squared error: 224.18216239510807
Mean squared error: 1433.161435648347
Mean squared error: 97.24766117212071
Mean squared error: 133.49096876834173
Mean squared error: 209.16763797720756
Mean squared error: 78.23726398522132
Mean squared error: 125.40655739993035
Mean squared error: 240.46431901946133
Mean squared error: 156.14505772688005
Mean squared error: 148.5061821351979
Mean squared error: 119.447315122542

[126.29349122419045,
 96.76143297806836,
 147.78868716177516,
 296.0136465608284,
 91.91169310247287,
 127.28017751071708,
 111.08590586565504,
 1045.128800130171,
 813.8786820929685,
 179.3520821927464,
 1183.900762301564,
 177.73368351466274,
 105.84584134997014,
 152.99098443251918,
 1266.5163162432736,
 224.18216239510807,
 1433.161435648347,
 97.24766117212071,
 133.49096876834173,
 209.16763797720756,
 78.23726398522132,
 125.40655739993035,
 240.46431901946133,
 156.14505772688005,
 148.5061821351979,
 119.4473151225429,
 99.63844957190675,
 157.38658191931367,
 580.293596323134,
 125.5075017468461,
 1830.3657767791467,
 155.6157196961666,
 1653.8798187348345,
 302.36657746708187,
 233.67593250571161,
 557.0317678715978,
 376.03814105361295,
 145.66496516902959,
 738.1506292474226,
 170.22941133979353,
 187.9324333183979,
 172.0389048662284,
 111.86477974353117,
 157.10336332825722,
 143.2944396350335,
 339.97054921318096,
 139.88501788943628,
 115.07271265036918,
 139.278983679

In [14]:
import numpy as np
StandardDeviation=np.std(mselist)
Mean=np.mean(mselist)
print("Mean is: ",Mean,"\nStandard Deviation is: ",StandardDeviation)

Mean is:  361.2094246813158 
Standard Deviation is:  426.48690673338723


### So the mean of MSE for Part A is 361.2 and STd is 426.48.
# -----------------------------------------------------------------------

## Part B

In [16]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
mselist=[]
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    mse=mean_squared_error(y_test, y_pred)
    mselist.append(mse)
    print('Mean squared error of epoch #',i+1," is:",mse)
    
StandardDeviation=np.std(mselist)
Mean=np.mean(mselist)
print("Mean is: ",Mean,"\nStandard Deviation is: ",StandardDeviation)

Mean squared error of epoch # 1  is: 466.0394893573828
Mean squared error of epoch # 2  is: 634.4509962897233
Mean squared error of epoch # 3  is: 406.10610377581446
Mean squared error of epoch # 4  is: 257.9501185259076
Mean squared error of epoch # 5  is: 294.3112752841028
Mean squared error of epoch # 6  is: 276.50373821697707
Mean squared error of epoch # 7  is: 279.3578397890933
Mean squared error of epoch # 8  is: 537.3595672037185
Mean squared error of epoch # 9  is: 381.14095504792886
Mean squared error of epoch # 10  is: 354.3396322250202
Mean squared error of epoch # 11  is: 229.68581606190642
Mean squared error of epoch # 12  is: 277.67998056145046
Mean squared error of epoch # 13  is: 317.0380019284259
Mean squared error of epoch # 14  is: 400.43565734007683
Mean squared error of epoch # 15  is: 323.55833566257627
Mean squared error of epoch # 16  is: 416.27510773467236
Mean squared error of epoch # 17  is: 405.6195149091279
Mean squared error of epoch # 18  is: 312.6800307

### So the mean of MSE for Part B is 362.98 and STd is 103.75. As compared to Part A, the mean almost remained the same but STd is considerbly very low. 
# -----------------------------------------------------------------------

## Part C

In [17]:
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=0)
    y_pred = model.predict(X_test)
    mse=mean_squared_error(y_test, y_pred)
    mselist.append(mse)
    print('Mean squared error of epoch #',i+1," is:",mse)
    
StandardDeviation=np.std(mselist)
Mean=np.mean(mselist)
print("Mean is: ",Mean,"\nStandard Deviation is: ",StandardDeviation)

Mean squared error of epoch # 1  is: 167.83500457421192
Mean squared error of epoch # 2  is: 156.40394927739908
Mean squared error of epoch # 3  is: 175.12899030713197
Mean squared error of epoch # 4  is: 144.01448685053415
Mean squared error of epoch # 5  is: 161.7845667245955
Mean squared error of epoch # 6  is: 141.39626884284527
Mean squared error of epoch # 7  is: 162.0415969793682
Mean squared error of epoch # 8  is: 143.40986685365368
Mean squared error of epoch # 9  is: 146.8487977659699
Mean squared error of epoch # 10  is: 148.2602162905278
Mean squared error of epoch # 11  is: 145.69056014717268
Mean squared error of epoch # 12  is: 150.7970727170423
Mean squared error of epoch # 13  is: 164.97175442391475
Mean squared error of epoch # 14  is: 184.44475153274777
Mean squared error of epoch # 15  is: 156.43370535657513
Mean squared error of epoch # 16  is: 154.25381070065887
Mean squared error of epoch # 17  is: 141.6430418447365
Mean squared error of epoch # 18  is: 162.4707

### So the mean of MSE for Part C is 258.73 and STd is 127.71. As compared to Part B, the mean is lower but STd is almost the same. This shows that the increasing the number of epochs the MSE decreases.  
# -----------------------------------------------------------------------

## Part D

In [18]:
def new_regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [20]:
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)
    model = new_regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    mse=mean_squared_error(y_test, y_pred)
    mselist.append(mse)
    print('Mean squared error of epoch #',i+1," is:",mse)
    
StandardDeviation=np.std(mselist)
Mean=np.mean(mselist)
print("Mean is: ",Mean,"\nStandard Deviation is: ",StandardDeviation)

Mean squared error of epoch # 1  is: 129.21012760157143
Mean squared error of epoch # 2  is: 88.91996169754194
Mean squared error of epoch # 3  is: 104.33099626303344
Mean squared error of epoch # 4  is: 126.58184489452307
Mean squared error of epoch # 5  is: 99.63102231128273
Mean squared error of epoch # 6  is: 109.38860174483129
Mean squared error of epoch # 7  is: 125.422371170726
Mean squared error of epoch # 8  is: 126.97239952539189
Mean squared error of epoch # 9  is: 121.00246885665565
Mean squared error of epoch # 10  is: 125.92643674479474
Mean squared error of epoch # 11  is: 90.3441880144527
Mean squared error of epoch # 12  is: 107.92734984967801
Mean squared error of epoch # 13  is: 110.85656967520957
Mean squared error of epoch # 14  is: 128.8964320108824
Mean squared error of epoch # 15  is: 125.2568484611665
Mean squared error of epoch # 16  is: 123.10399710276955
Mean squared error of epoch # 17  is: 127.56672083997188
Mean squared error of epoch # 18  is: 135.977736

### So the mean of MSE for Part D is 202.6 and STd is 159.1. As compared to Part B, the mean is considrbly low but Std is bit higher which shows that the MSE in Part D are more spreaded away from the mean. 
# -----------------------------------------------------------------------

# Thank you!